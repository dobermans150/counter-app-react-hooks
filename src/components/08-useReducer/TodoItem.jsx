import React from 'react'
import PropTypes from 'prop-types'

export const TodoItem = ( { todo, index, handleToggle, handleDelete } ) => {
    return (
        <li className="list-group-item" key={ todo.id }>
            <p
                className={ `${todo.done && 'complete'}` }
                onClick={ () => handleToggle( todo.id ) }

            >
                { index + 1 }. { todo.desc }
            </p>
            <button
                className="btn btn-danger"
                onClick={ () => handleDelete( todo.id ) }
            >
                Borrar
            </button>
        </li>
    )
}





TodoItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}

