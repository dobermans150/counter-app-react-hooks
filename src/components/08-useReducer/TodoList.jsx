import React from 'react'
import PropTypes from 'prop-types'

import { TodoItem } from './TodoItem'

export const TodoList = ( { todos, handleToggle, handleDelete } ) => {


    return (
        <ol className="list-group list-group-flush">
            { todos.map( ( todo, i ) => (
                <TodoItem key={ todo.id } todo={ todo } index={ i } handleDelete={ handleDelete } handleToggle={ handleToggle } />
            ) ) }
        </ol>
    )
}





TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}


