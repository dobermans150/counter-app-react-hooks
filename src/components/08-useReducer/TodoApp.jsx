import React, { useEffect, useReducer } from 'react';

import './Styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse( localStorage.getItem( 'todos' ) ) || [];
};

export const TodoApp = () => {
    const [ todos, dispatch ] = useReducer( todoReducer, [], init );


    useEffect( () => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) );
    }, [ todos ] );

    const handleDelete = ( todoId ) => {
        const deleteItem = {
            type: 'delete',
            payload: todoId,
        };

        dispatch( deleteItem );
    };

    const handleToggle = ( todoId ) => {
        dispatch( { type: 'toggle', payload: todoId } );
    };

    const handleAddTodo = ( newTodo ) => {

        dispatch( {
            type: 'add',
            payload: newTodo,
        } );
    }


    return (
        <div>
            <h1>ToDo App ( { todos.length } ) </h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList todos={ todos } handleToggle={ handleToggle } handleDelete={ handleDelete } />
                </div>
                <div className="col">
                    <TodoAdd handleAddTodo={ handleAddTodo } />
                </div>
            </div>
        </div>
    );
};
