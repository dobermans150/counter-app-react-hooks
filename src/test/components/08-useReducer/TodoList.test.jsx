import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixture/demotodos';

describe( 'Pruebas en el componente TodoList <TodoList  />', () => {
    const todos = demoTodos;
    const handleToggle = jest.fn();
    const handleDelete = jest.fn();


    let wrapper = shallow(
        <TodoList
            todos={ todos }
            handleToggle={ handleToggle }
            handleDelete={ handleDelete }
        />
    )
    beforeEach( () => {
        wrapper = shallow(
            <TodoList
                todos={ todos }
                handleToggle={ handleToggle }
                handleDelete={ handleDelete }
            />
        )
    } )

    test( 'Debe de mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot()

    } )


    test( 'debe tener n cantidades de <TodoItem  />', () => {

        expect( wrapper.find( 'TodoItem' ).length ).toBe( demoTodos.length )
        expect( wrapper.find( 'TodoItem' ).at( 0 ).prop( 'handleDelete' ) ).toEqual( expect.any( Function ) )
    } )


} )
