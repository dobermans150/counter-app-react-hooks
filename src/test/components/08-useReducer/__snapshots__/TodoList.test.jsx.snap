// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente TodoList <TodoList  /> Debe de mostrarse correctamente 1`] = `
<ol
  className="list-group list-group-flush"
>
  <TodoItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ol>
`;
